### DISCUSSION TITLE
Anyone know of a good html minifier that works well with cotton?

### DISCUSSION THREAD

**Discussion by ra-dave** [wrabit/django-cotton#184] [Q&A]
I'm on the search for a minifier to, at a minimum, remove whitespace and comments from the template .html files.

**Comment by wrabit** [comment]
~~@ra-dave did you look at https://github.com/cobrateam/django-htmlmin ? Seems like a decent choice to try.~~

Issues are somewhat stale. 

Try https://github.com/adamchainz/django-minify-html instead, it's more active.

**Comment by ra-dave** [comment]
I've evaluated all the top packages and haven't found one yet that doesn't have an issue because of django's complex templating.  At least I found a decent inline js compressor (django-compressor).

Would still like to find a good whitespace reduction solution as that would achieve a high level of reduction while (presumably) being pretty safe.

**Comment by mattbha** [comment]
Can you elaborate on the desire/usecase/benchmarks? In my experience I ignore minification and just let gzip handle compression, which has always seemed to work fine and be completely stable. 

I think adding minification might optimize bytes over the network slightly, while adding a little extra CPU server-side. And of course an additional complexity in the pipeline via the plugin itself.

I **could** see comment removal being useful. That would allow us to put more comments in our templates without fear of it appearing in the source -- but even a component that is repeated 100 times will have any comments compressed with gzip.

Interesting idea

**Comment by mattbha** [comment]
Keep in mind, if you structure your comments

`{# like this #}` django will automatically remove them during rendering.

**Comment by sveetch** [comment]
Minifier are good to use on rendered content and for this the Django builtin tag "spaceless" is doing a nice work while incomplete since it only remove whitespaces between tags and ignore every other ones. 

Commonly you just put the tag once in the higher template and it will remove space from all the lower template contents. Just be aware it can break layout for a code preview (inside a `<pre>` like if you display some code with Pygments) where you would just have to apply some config to transform whitespace into Non-breaking space (`&nbsp;`).

However if you just want prettify your templates, you have [djLint](https://github.com/djlint/djLint) that does work well and it is also a linter. It has some opinionated behaviors but commonly you can bypass them with some configuration.
